File: server/functions/src/models/domain/possession_state.ts
import { Income } from './income';
import { Expense } from './expense';
import { Asset } from './asset';
import { Liability } from './liability';
import { removeKeys } from '../../utils/object';

export interface PossessionState {
  readonly incomes: Income[];
  readonly expenses: Expense[];
  readonly assets: Asset[];
  readonly liabilities: Liability[];
}

export namespace PossessionStateEntity {
  export const normalize = (possessionState: PossessionState) => {
    const newState = JSON.parse(JSON.stringify(possessionState)) as PossessionState;
    removeKeys(newState, ['id']);
    return newState;
  };

  export const createEmpty = (): PossessionState => ({
    incomes: [],
    expenses: [],
    assets: [],
    liabilities: []
  });
}

File: client/lib/models/domain/game/possession_state/possession_state.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'assets/asset.dart';
import 'expenses/expense.dart';
import 'incomes/income.dart';
import 'liabilities/liability.dart';

part 'possession_state.freezed.dart';
part 'possession_state.g.dart';

@freezed
class PossessionState with _$PossessionState {
  factory PossessionState({
    @JsonKey(defaultValue: []) required List<Income> incomes,
    @JsonKey(defaultValue: []) required List<Expense> expenses,
    @JsonKey(defaultValue: []) required List<Asset> assets,
    @JsonKey(defaultValue: []) required List<Liability> liabilities,
  }) = _PossessionState;

  factory PossessionState.fromJson(Map<String, dynamic> json) =>
      _$PossessionStateFromJson(json);
}
